### ISSUE TITLE
Is it possible to use levenshtein as the scorer for extractOne?

### ISSUE THREAD

**Issue by kdi-insitro** [rapidfuzz/RapidFuzz#89]
Given a string and a list of strings I would like to find the best match that is < some max distance using Levenshtein distance. I think `extractOne` is what I'm looking for, however, when trying to use `extractOne` with `levenshtein` as the scorer I get the error `TypeError: levenshtein() got an unexpected keyword argument 'processor'`. I assume this is because `levenshtein` doesn't take a `processor` argument.

In my case, all my strings are preprocessed to all caps so I don't need to use a processor but still get the same error when using `processor = None` 

To get around this I'm doing something like:
```
Dist = namedtuple("Dist", ["dist", "query", "choice"])

def best_match(query, choices, max_dist):
    min_dist = max_dist
    best_match = None
    for choice in choices:
        dist = string_metric.levenshtein(query,choice, max=max_dist)
        if dist == 0:
            best_match = Dist(dist, query, choice)
            break
        if dist != -1 and dist < min_dist:
            best_match = Dist(dist, query, choice)
            min_dist = dist
    return best_match
```
I would guess this is slower than `extractOne` but I'm not sure how to accomplish this otherwise.

**Comment by maxbachmann** [comment]
The reason I did not add this yet, is that it is kind of the opposite behavior of extractOne. extractOne calls scorers and returns the one with the highest score. For distances like `levenshtein` or `hamming` I would need to return the one with the lowest score. For scorers/distances that are part of RapidFuzz this could be solved in a simple way, by using different implementations for the two cases. However I was unsure how I should treat custom functions by the user, since I do not know which way they work:
```python
def custom_scorer(a,b,**kwargs):
    return 1.0

process.extractOne(a, b, scorer=custom_scorer)
``` 
also `score_cutoff` would be `max` for the distances.

I am thinking about keeping extract/extractOne around for compatibility with FuzzyWuzzy and adding new processors for other use cases like e.g. matching lists against lists. One of the main reasons I did not start adding them already is, that I am still thinking about the naming convention + arguments. Essentially I would like to provide the following things:

1) there is still a open issue on functions to match two lists to a matrix. The idea is to implement this similar to cdist/pdist in scipy, so it returns a numpy matrix with scores
2) There should be an alternative to extract, that returns a numpy array with scores instead of the current solution with limit, that returns a complex list. This has the following reasons:
  - extract is very slow when returning all scorers, simply because the creation of a list filled with tuples takes a big amount of the executions time
  - numpy arrays are a lot better to handle when working with big amounts of data without taking up a ton of memory
3) as you mentioned there should be an alternative to extractOne, that works both for distances and normalized distances.
   The main open question here is how it should handle custom functions by a users, since we do not know whether it is a distance or normalized distance (extra argument max/min as a metric to find the best score?).  Performance for custom scorers is a lot less critical. They are called through Python anyways, so it does not help with performance, so it is mostly there to make it easier to code
4) all processors should have multi threaded alternatives in a multiprocess module. As a difference it would not take any custom scorers (possibly no custom processors either), since multi threading requires the GIL to be releases. (for custom functions the only option would be a multiprocess, which I would at least in the start not provide)

I am still not quite sure on the following points:
- naming conventions of these functions
- the arguments they take, so they are somewhat consistent across all the functions and allow the use of both distances and normalized_distances

Ideas are welcome ;)


P.S. how much slower your Python function is, is largely dependent on the string lengths your working on. For short strings it can take a lot longer, while for very long strings (multiple hundred chars) the Levenshtein calculation takes a while and so the Python overhead is less of an issue
